File: frontend/src/services/roomService.js
import api from './api';

export const roomService = {
  getRooms: async (filters = {}) => {
    try {
      console.log('[RoomService] Getting rooms with filters:', filters);

      // Формируем параметры запроса
      const params = new URLSearchParams();
      if (filters.type) params.append('room_type', filters.type);
      if (filters.status) params.append('status', filters.status);

      const url = params.toString() ? `rooms?${params}` : 'rooms';
      const response = await api.get(url);

      console.log('[RoomService] API Response:', response.data);
      console.log('[RoomService] Response type:', typeof response.data);
      console.log('[RoomService] Is array:', Array.isArray(response.data));

      // Проверяем что получили массив
      if (!Array.isArray(response.data)) {
        console.error('[RoomService] Response is not an array:', response.data);
        return [];
      }

      // Маппинг типов комнат для отображения
      const roomTypeMap = {
        "2 o'rinli standart": "2 o'rinli standart",
        "4 o'rinli standart": "4 o'rinli standart",
        "2 o'rinli lyuks": "2 o'rinli lyuks",
        "4 o'rinli kichik VIP": "4 o'rinli kichik VIP",
        "4 o'rinli katta VIP": "4 o'rinli katta VIP",
        "4 o'rinli apartament": "4 o'rinli apartament",
        "Kottedj (6 kishi uchun)": "Kottedj (6 kishi uchun)",
        "Prezident apartamenti (8 kishi uchun)": "Prezident apartamenti (8 kishi uchun)",
        // Альтернативные названия на случай если backend вернет другие
        'STANDARD_DOUBLE': "2 o'rinli standart",
        'STANDARD_QUAD': "4 o'rinli standart",
        'LUX_DOUBLE': "2 o'rinli lyuks",
        'VIP_SMALL': "4 o'rinli kichik VIP",
        'VIP_LARGE': "4 o'rinli katta VIP",
        'APARTMENT': "4 o'rinli apartament",
        'COTTAGE': "Kottedj (6 kishi uchun)",
        'PRESIDENT': "Prezident apartamenti (8 kishi uchun)"
      };

      // Преобразуем данные
      const rooms = response.data.map(room => ({
        ...room,
        room_type: roomTypeMap[room.room_type] || room.room_type,
        // Убеждаемся что все поля есть
        capacity: room.capacity || 2,
        price_per_night: room.price_per_night || 500000,
        description: room.description || '',
        amenities: room.amenities || '',
        is_available: room.is_available !== undefined ? room.is_available : true,
        current_booking: room.current_booking || null
      }));

      console.log('[RoomService] Processed rooms count:', rooms.length);
      console.log('[RoomService] First 3 rooms:', rooms.slice(0, 3));

      return rooms;
    } catch (error) {
      console.error('[RoomService] getRooms error:', error);
      console.error('[RoomService] Error response:', error.response);

      // Возвращаем пустой массив при ошибке
      return [];
    }
  },

  getRoom: async (roomId) => {
    try {
      console.log('[RoomService] Getting room:', roomId);
      const response = await api.get(`rooms/${roomId}`);
      console.log('[RoomService] Room data:', response.data);
      return response.data;
    } catch (error) {
      console.error('[RoomService] getRoom error:', error);
      throw error;
    }
  },

  updateRoom: async (roomId, data) => {
    try {
      console.log('[RoomService] Updating room:', roomId, data);
      const response = await api.patch(`rooms/${roomId}`, data);
      console.log('[RoomService] Room updated:', response.data);
      return response.data;
    } catch (error) {
      console.error('[RoomService] updateRoom error:', error);
      throw error;
    }
  },

  checkAvailability: async (roomId, startDate, endDate) => {
    try {
      const params = new URLSearchParams({
        start_date: startDate,
        end_date: endDate
      });
      const response = await api.get(`rooms/${roomId}/availability?${params}`);
      console.log('[RoomService] Availability:', response.data);
      return response.data;
    } catch (error) {
      console.error('[RoomService] checkAvailability error:', error);
      throw error;
    }
  }
};

File: backend/app/services/room_service.py
from sqlalchemy.orm import Session
from typing import List, Optional
from datetime import date
from ..models.room import Room
from ..models.booking import Booking


class RoomService:
    @staticmethod
    def get_rooms_with_status(
            db: Session,
            room_type_filter: Optional[str] = None,
            status_filter: Optional[str] = None
    ) -> List[dict]:
        """
        Получает список комнат с их текущим статусом бронирования и применяет фильтры.
        Это основной метод, который будет использовать API.
        """
        # Начинаем базовый запрос для всех комнат
        query = db.query(Room)

        # Применяем фильтр по типу комнаты, если он указан
        if room_type_filter:
            query = query.filter(Room.room_type == room_type_filter)

        all_rooms = query.order_by(Room.id).all()

        # Получаем ID всех комнат, которые заняты СЕГОДНЯ
        today = date.today()
        occupied_room_ids_query = db.query(Booking.room_id).filter(
            Booking.start_date <= today,
            Booking.end_date > today  # > чтобы не считать день выезда
        ).distinct()

        # Преобразуем в set для быстрой проверки (например, `if room_id in occupied_ids`)
        occupied_ids = {room_id for (room_id,) in occupied_room_ids_query}

        result_rooms = []
        for room in all_rooms:
            is_occupied = room.id in occupied_ids

            # Применяем фильтр по статусу, если он указан
            if status_filter:
                if status_filter == "available" and is_occupied:
                    continue  # Пропускаем занятые, если нужен "свободен"
                elif status_filter == "occupied" and not is_occupied:
                    continue  # Пропускаем свободные, если нужен "занят"

            # Добавляем в результат. Pydantic-схема сама преобразует это в нужный JSON.
            # Мы добавляем поле is_available "на лету".
            room.is_available = not is_occupied
            result_rooms.append(room)

        return result_rooms

    @staticmethod
    def get_room_with_status(db: Session, room_id: int) -> Optional[Room]:
        """
        Получает одну комнату по ID и добавляет её текущий статус.
        """
        room = db.query(Room).filter(Room.id == room_id).first()
        if not room:
            return None

        today = date.today()
        # Проверяем, занята ли конкретно эта комната
        is_occupied = db.query(Booking).filter(
            Booking.room_id == room_id,
            Booking.start_date <= today,
            Booking.end_date > today
        ).first() is not None

        room.is_available = not is_occupied
        return room

    # --- Остальные ваши методы остаются без изменений ---

    @staticmethod
    def initialize_rooms(db: Session):
        """Initialize rooms with default data if empty"""
        if db.query(Room).count() > 0:
            return

        rooms_data = [
            {"room_number": "101", "room_type": "2 o'rinli standart", "capacity": 2, "price_per_night": 500000},
            {"room_number": "102", "room_type": "2 o'rinli standart", "capacity": 2, "price_per_night": 500000},
            {"room_number": "103", "room_type": "2 o'rinli standart", "capacity": 2, "price_per_night": 500000},
            {"room_number": "104", "room_type": "2 o'rinli standart", "capacity": 2, "price_per_night": 500000},
            {"room_number": "105", "room_type": "2 o'rinli standart", "capacity": 2, "price_per_night": 500000},
            {"room_number": "106", "room_type": "2 o'rinli standart", "capacity": 2, "price_per_night": 500000},
            {"room_number": "107", "room_type": "2 o'rinli standart", "capacity": 2, "price_per_night": 500000},
            {"room_number": "108", "room_type": "2 o'rinli standart", "capacity": 2, "price_per_night": 500000},
            {"room_number": "109", "room_type": "2 o'rinli standart", "capacity": 2, "price_per_night": 500000},
            {"room_number": "110", "room_type": "2 o'rinli standart", "capacity": 2, "price_per_night": 500000},
            {"room_number": "201", "room_type": "4 o'rinli standart", "capacity": 4, "price_per_night": 700000},
            {"room_number": "202", "room_type": "4 o'rinli standart", "capacity": 4, "price_per_night": 700000},
            {"room_number": "203", "room_type": "4 o'rinli standart", "capacity": 4, "price_per_night": 700000},
            {"room_number": "204", "room_type": "4 o'rinli standart", "capacity": 4, "price_per_night": 700000},
            {"room_number": "205", "room_type": "4 o'rinli standart", "capacity": 4, "price_per_night": 700000},
            {"room_number": "206", "room_type": "4 o'rinli standart", "capacity": 4, "price_per_night": 700000},
            {"room_number": "301", "room_type": "2 o'rinli lyuks", "capacity": 2, "price_per_night": 800000},
            {"room_number": "302", "room_type": "2 o'rinli lyuks", "capacity": 2, "price_per_night": 800000},
            {"room_number": "303", "room_type": "2 o'rinli lyuks", "capacity": 2, "price_per_night": 800000},
            {"room_number": "304", "room_type": "2 o'rinli lyuks", "capacity": 2, "price_per_night": 800000},
            {"room_number": "305", "room_type": "2 o'rinli lyuks", "capacity": 2, "price_per_night": 800000},
            {"room_number": "401", "room_type": "4 o'rinli kichik VIP", "capacity": 4, "price_per_night": 1000000},
            {"room_number": "402", "room_type": "4 o'rinli kichik VIP", "capacity": 4, "price_per_night": 1000000},
            {"room_number": "403", "room_type": "4 o'rinli kichik VIP", "capacity": 4, "price_per_night": 1000000},
            {"room_number": "404", "room_type": "4 o'rinli kichik VIP", "capacity": 4, "price_per_night": 1000000},
            {"room_number": "501", "room_type": "4 o'rinli katta VIP", "capacity": 4, "price_per_night": 1200000},
            {"room_number": "502", "room_type": "4 o'rinli katta VIP", "capacity": 4, "price_per_night": 1200000},
            {"room_number": "503", "room_type": "4 o'rinli katta VIP", "capacity": 4, "price_per_night": 1200000},
            {"room_number": "601", "room_type": "4 o'rinli apartament", "capacity": 4, "price_per_night": 1500000},
            {"room_number": "602", "room_type": "4 o'rinli apartament", "capacity": 4, "price_per_night": 1500000},
            {"room_number": "701", "room_type": "Kottedj (6 kishi uchun)", "capacity": 6, "price_per_night": 2000000},
            {"room_number": "801", "room_type": "Prezident apartamenti (8 kishi uchun)", "capacity": 8,
             "price_per_night": 3000000},
        ]
        for room_data in rooms_data:
            db.add(Room(**room_data))
        db.commit()

    @staticmethod
    def get_room(db: Session, room_id: int) -> Optional[Room]:
        return db.query(Room).filter(Room.id == room_id).first()